cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -339,18 +340,48 @@ fn handle_vaa(deps: DepsMut<WormholeQuery>, vaa: Binary) -> anyhow::Result<Event
             .into(),
         )
         .context(ContractError::VerifyQuorum)?;
+
+    let digest = vaa::digest(data)
+        .map(|d| d.secp256k_hash.to_vec().into())
+        .context("failed to calculate digest for VAA body")?;
+
     let (body, payload) = serde_wormhole::from_slice_with_payload::<Body<()>>(data)
         .context("failed to parse VAA body")?;
 
-    if body.emitter_chain == Chain::Solana && body.emitter_address == wormhole::GOVERNANCE_EMITTER {
+    let digest_key = DIGESTS.key((
+        body.emitter_chain.into(),
+        body.emitter_address.0.to_vec(),
+        body.sequence,
+    ));
+
+    if let Some(saved_digest) = digest_key
+        .may_load(deps.storage)
+        .context("failed to load transfer digest")?
+    {
+        if saved_digest != digest {
+            bail!(ContractError::DigestMismatch);
+        }
+
+        bail!(ContractError::DuplicateMessage);
+    }
+
+    let evt = if body.emitter_chain == Chain::Solana
+        && body.emitter_address == wormhole::GOVERNANCE_EMITTER
+    {
         let govpacket =
             serde_wormhole::from_slice(payload).context("failed to parse governance packet")?;
-        handle_governance_vaa(deps, body.with_payload(govpacket))
+        handle_governance_vaa(deps.branch(), body.with_payload(govpacket))?
     } else {
         let (msg, _) = serde_wormhole::from_slice_with_payload(payload)
             .context("failed to parse tokenbridge message")?;
-        handle_tokenbridge_vaa(deps, body.with_payload(msg))
-    }
+        handle_tokenbridge_vaa(deps.branch(), body.with_payload(msg))?
+    };
+
+    digest_key
+        .save(deps.storage, &digest)
+        .context("failed to save message digest")?;
+
+    Ok(evt)
 }
 
 fn handle_governance_vaa(
```

Comment:
I think we should save the governance VAAs in `DIGESTS` like how you were saving the hashes in the previous `GOVERNANCE_VAAS` storage so none of the previous governance actions can be replayed.

Reply:
This was moved to lines 380-382 so that saving the digests is shared between the `handle_governance_vaa` and `handle_tokenbridge_vaa` methods.  So they are still being saved.  There is also a test for this: https://github.com/wormhole-foundation/wormhole/blob/acc3ec14d8264075a6c8d084d007538d3def395d/cosmwasm/contracts/wormchain-accounting/tests/chain_registration.rs#L122-L135

Reply:
Doh! That's right, looks good